magma/value_utils.py

```diff
@@ -0,0 +1,107 @@
+import dataclasses
+import typing
+
+from magma.ref import Ref, TupleRef, ArrayRef
+from magma.type_utils import isprotocol, isdigital, isbits, isarray, istuple
+
+
+def _unwrap_protocol_value(value):
+    T = type(value)
+    if isprotocol(T):
+        return value._get_magma_value_()
+    return value
+
+
+@dataclasses.dataclass(frozen=True)
+class _ValueWrapper:
+    raw: typing.Any
+    children: typing.List[typing.Any]
+    name: str
+
+
+def _wrap_value(value):
+    if isinstance(value, _ValueWrapper):
+        return value
+    value = _unwrap_protocol_value(value)
+    T = type(value)
+
+    if isdigital(T):
+        return _ValueWrapper(value, [], "Digital")
+    if isbits(T):
+        return _ValueWrapper(value, [], "Bits")
+    if isarray(T):
+        return _ValueWrapper(value, [t for t in value], "Array")
+    if istuple(T):
+        return _ValueWrapper(value, [t for t in value], "Tuple")
+
+    raise NotImplementedError(T)
+
+
+class ValueVisitor:
+    def visit(self, value):
+        wrapped = _wrap_value(value)
+        method = 'visit_' + wrapped.name
+        visitor = getattr(self, method, self.generic_visit)
+        return visitor(value)
```

Comment:
Hmm, this seems a little inconsistent with the TypeVisitor pattern, but not sure if this is intended.

Here we visit with wrapped value, while in TypeVisitor visits with the raw value.  The main potential wonkiness I see is that if we use generic_visit here, we'll wrap twice, not sure if that's intended?

Reply:
It's not clear because of the code but notice that here we still call `visitor(value)` (not `visitor(wrapped)`). So the code paths are the same - I should've used `wrapped.raw` here to be more consistent. Actually, I've pulled this all into a common class so it should be more clear now.